### PR TITLE
[resotolib][feat] Implement a dependent node iterator

### PR DESCRIPTION
**Pull request by aquamatthias**
# Description
Walk a loosely connected graph by dependent edges.
Emit list of nodes in generations, so one generation of nodes always only depend on previous generations of nodes.

# To-Dos

- [x] Add test coverage for new or updated functionality
- [x] Lint and test with `tox`

# Issues Fixed

- Fixes #XXXX

# Code of Conduct

By submitting this pull request, I agree to follow the [code of conduct](https://resoto.com/code-of-conduct).

### CODE REVIEW

**Review by meln1k — APPROVED** [review]
LGTM

File: resotolib/test/test_graph_extensions.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    result = list(dependent_node_iterator(graph))
+    assert len(result) == 3  # 3 steps to complete
+    assert result == [
+        nodes(3, 5, 7, 10, 11, 13),  # step 1
+        nodes(2, 6, 9, 12),  # step 2
+        nodes(1, 4, 8),  # step 3
+    ]
+
+
+def test_delete_nodes(graph: DiGraph):
```
Comment (meln1k) [inline resotolib/test/test_graph_extensions.py:35]: Do you also want to cover the edge cases such as an empty graph?

Comment (aquamatthias) [reply resotolib/test/test_graph_extensions.py:35]: Added a test case.

**Review by meln1k — APPROVED** [review]

**Review by lloesche — APPROVED** [review]
lgtm, thank you for your help! 🚢
